### ISSUE TITLE
NioChannelMemoryLeakTest.testNioChannelLeak_afterMultipleSplitBrainMerges

### ISSUE THREAD

**Issue by tezc** [hazelcast/hazelcast#14985]
_master_ (commit d2170c24553ab21e64f14a4eccbf156397d08431)

Failed on Windows : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-Windows-OracleJDK8/38/testReport/com.hazelcast.internal.networking.nio/NioChannelMemoryLeakTest/testNioChannelLeak_afterMultipleSplitBrainMerges/

**Stacktrace**
```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.assertNoChannelLeak(NioChannelMemoryLeakTest.java:116)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.access$000(NioChannelMemoryLeakTest.java:46)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest$2.run(NioChannelMemoryLeakTest.java:98)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1310)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1412)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.testNioChannelLeak_afterMultipleSplitBrainMerges(NioChannelMemoryLeakTest.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Standard output :** 

```
Hiccups measured while running test 'testNioChannelLeak_afterMultipleSplitBrainMerges(com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest):'
03:37:20, accumulated pauses: 732 ms, max pause: 8 ms, pauses over 1000 ms: 0
03:37:25, accumulated pauses: 733 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:37:30, accumulated pauses: 723 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:37:35, accumulated pauses: 726 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:37:40, accumulated pauses: 732 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:37:45, accumulated pauses: 728 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:37:50, accumulated pauses: 727 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:37:55, accumulated pauses: 734 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:38:00, accumulated pauses: 728 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:38:05, accumulated pauses: 720 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:10, accumulated pauses: 725 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:15, accumulated pauses: 742 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:20, accumulated pauses: 746 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:25, accumulated pauses: 737 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:30, accumulated pauses: 725 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:35, accumulated pauses: 716 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:38:40, accumulated pauses: 720 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:38:45, accumulated pauses: 719 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:38:50, accumulated pauses: 735 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:38:55, accumulated pauses: 740 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:00, accumulated pauses: 724 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:05, accumulated pauses: 728 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:10, accumulated pauses: 727 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:39:15, accumulated pauses: 718 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:20, accumulated pauses: 727 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:25, accumulated pauses: 732 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:30, accumulated pauses: 714 ms, max pause: 1 ms, pauses over 1000 ms: 0
03:39:35, accumulated pauses: 999 ms, max pause: 311 ms, pauses over 1000 ms: 0
03:39:40, accumulated pauses: 719 ms, max pause: 2 ms, pauses over 1000 ms: 0
03:39:45, accumulated pauses: 274 ms, max pause: 1 ms, pauses over 1000 ms: 0
```

**Comment by vbekiaris** [comment]
Couldn't reproduce locally, after thousands of executions. #15047 adds logging to the test so we can find out which connection is leftover on which instance and identify the root cause when the test fails again.

**Comment by tezc** [comment]
http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.maintenance-Windows-OracleJDK8/106/testReport/com.hazelcast.internal.networking.nio/NioChannelMemoryLeakTest/testNioChannelLeak_afterMultipleSplitBrainMerges/

**Comment by vojtechtoman** [comment]
For some reason, the local socket address of member 3 is always 0.0.0.0 (and not 127.0.0.1 like the first two members) in both failed runs. Not sure if it's significant here, just mentioning it here.

**Comment by mmedenjak** [comment]
Closing with additional logging until next failure. Thank you for the change @vojtechtoman !

**Comment by vbekiaris** [comment]
Test failed in 3.12.z branch PR builder job http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/213:

```
java.lang.AssertionError
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.assertNoChannelLeak(NioChannelMemoryLeakTest.java:116)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.access$000(NioChannelMemoryLeakTest.java:46)
	at com.hazelcast.internal.networking.nio.NioChannelMemoryLeakTest.testNioChannelLeak_afterMultipleSplitBrainMerges(NioChannelMemoryLeakTest.java:93)
```

Unfortunately the additional logging is only available in 4.0+ branches
